src/deepsparse/timing/timing_builder.py

```diff
@@ -0,0 +1,197 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+from deepsparse.timing.timing_schema import InferenceTimingSchema
+
+
+__all__ = ["TimingBuilder"]
+
+
+class TimingBuilder:
+    """
+    The goal of the object is to aggregate the durations
+    (time deltas in seconds) of various components of the inference
+    pipeline.
+    Once all the desired time deltas are aggregated, the
+    object may build the data structure that summarizes the
+    collected information.
+
+    The envisioned flow of how the object aggregates time
+    deltas is depicted below
+
+    ```
+    builder.start()
+
+    builder.{process_name}_start()
+    something_happens()
+    builder.{process_name}_complete()
+    ...
+    summary = builder.build()
+    ```
+
+    The object can be extended to aggregate further deltas.
+    To extend, four things are required:
+        - adding {process_name}_delta property (including the setter)
+        - adding {process_name}_start method
+        - adding {process_name}_complete method
+        - adding the {process_name_delta} to the arguments collected
+          inside the build() method.
+    """
+
+    def __init__(self):
+        self.started = False
+
+        # general placeholder for start time
+        self._t0 = None
+        # generate placeholder for completion time
+        self._t1 = None
```

Comment:
why do we need this? looks like it adds the potential for annoying bugs if the builder methods are called out of order for some reason.  see longer comment above about refactoring, shouldn't have a need for this after.